File: web/src/models/account.ts
import { z } from "zod";

import { userSchema, type User } from "@/models/user";

export interface Account {
  id: string;
  user_id: string;
  provider: string;
  external_id: string;
  created_at: Date;
  updated_at: Date;
  user?: User | null;
}

export const accountSchema: z.ZodType<Account> = z.lazy(() =>
  z
    .object({
      id: z.uuid(),
      user_id: z.uuid(),
      provider: z.string(),
      external_id: z.string(),
      created_at: z.coerce.date(),
      updated_at: z.coerce.date(),
      user: userSchema.nullable().optional(),
    })
    .strict(),
);

File: internal/model/account.go
package model

import "time"

type Account struct {
	ID string `gorm:"type:uuid;primaryKey;default:gen_random_uuid()" json:"id"`

	UserID     string `gorm:"type:uuid;not null;index" json:"user_id"`
	Provider   string `gorm:"type:varchar(255);not null;uniqueIndex:idx_accounts_provider_external_id,priority:1" json:"provider"`
	ExternalID string `gorm:"type:varchar(255);not null;uniqueIndex:idx_accounts_provider_external_id,priority:2" json:"external_id"`

	CreatedAt time.Time `gorm:"type:timestamp;not null;default:now()" json:"created_at"`
	UpdatedAt time.Time `gorm:"type:timestamp;not null;default:now()" json:"updated_at"`

	User *User `gorm:"foreignKey:UserID;references:ID;constraint:OnDelete:CASCADE" json:"user"`
}
